From c83121a603bc2caa02217a0834a5f5368ca34da2 Mon Sep 17 00:00:00 2001
From: Paul Harrison <pharrison@mythtv.org>
Date: Sat, 8 Feb 2014 19:48:51 +0000
Subject: [PATCH 3/6] MusicMetadata: add a getLocalFilename() function

This will attempt to find the file on the local file system either using the
filename as is or it will also look in the local 'Music' storage group.
---
 mythtv/libs/libmythmetadata/musicmetadata.cpp |   27 ++++++++++++++++++++++++-
 mythtv/libs/libmythmetadata/musicmetadata.h   |    1 +
 2 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/mythtv/libs/libmythmetadata/musicmetadata.cpp b/mythtv/libs/libmythmetadata/musicmetadata.cpp
index 9e95070..041cacf 100644
--- a/mythtv/libs/libmythmetadata/musicmetadata.cpp
+++ b/mythtv/libs/libmythmetadata/musicmetadata.cpp
@@ -15,6 +15,7 @@
 #include <mythlogging.h>
 #include <mythdate.h>
 #include <remotefile.h>
+#include <storagegroup.h>
 
 // libmythmetadata
 #include "musicmetadata.h"
@@ -763,6 +764,18 @@ QString MusicMetadata::Filename(bool find)
     return QString();
 }
 
+/// try to find the track on the local file system
+QString MusicMetadata::getLocalFilename(void)
+{
+    // try the file name as is first
+    if (QFile::exists(m_filename))
+        return m_filename;
+
+    // not found so now try to find the file in the local 'Music' storage group
+    StorageGroup storageGroup("Music", gCoreContext->GetHostName(), false);
+    return storageGroup.FindFile(m_filename);
+}
+
 void MusicMetadata::setField(const QString &field, const QString &data)
 {
     if (field == "artist")
@@ -1070,7 +1083,19 @@ void MusicMetadata::reloadAlbumArtImages(void)
 // NOTE the caller is responsible for deleting it
 MetaIO* MusicMetadata::getTagger(void)
 {
-    return MetaIO::createTagger(Filename(true));
+    // the taggers require direct file access so try to find
+    // the file on the local filesystem
+
+    QString filename = getLocalFilename();
+
+    if (!filename.isEmpty())
+    {
+        LOG(VB_FILE, LOG_INFO, QString("MusicMetadata::getTagger - creating tagger for %1").arg(filename));
+        return MetaIO::createTagger(filename);
+    }
+
+    LOG(VB_GENERAL, LOG_ERR, QString("MusicMetadata::getTagger - failed to find %1 on the local filesystem").arg(Filename(false)));
+    return NULL;
 }
 
 //--------------------------------------------------------------------------
diff --git a/mythtv/libs/libmythmetadata/musicmetadata.h b/mythtv/libs/libmythmetadata/musicmetadata.h
index 378a15a..72a37c1 100644
--- a/mythtv/libs/libmythmetadata/musicmetadata.h
+++ b/mythtv/libs/libmythmetadata/musicmetadata.h
@@ -198,6 +198,7 @@ class META_PUBLIC MusicMetadata
 
     QString Filename(bool find = true);
     void setFilename(const QString &lfilename) { m_filename = lfilename; }
+    QString getLocalFilename(void);
 
     QString Hostname(void) { return m_hostname; }
     void setHostname(const QString &host) { m_hostname = host; }
-- 
1.7.10

